### PR TITLE
[SPARK-11416][BUILD] Update to Chill 0.8.0 & Kryo 3.0.3

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch upgrades Chill to 0.8.0 and Kryo to 3.0.3. While we'll likely need to bump these dependencies again before Spark 2.0 (due to SPARK-14221 / https://github.com/twitter/chill/issues/252), I wanted to get the bulk of the Kryo 2 -> Kryo 3 migration done now in order to figure out whether there are any unexpected surprises.

### CODE REVIEW

File: dev/deps/spark-deps-hadoop-2.2 [hunk 1]
```diff
@@ -122,7 +122,7 @@ jsr305-1.3.9.jar
 jta-1.1.jar
 jtransforms-2.4.0.jar
 jul-to-slf4j-1.7.16.jar
-kryo-2.21.jar
+kryo-shaded-3.0.3.jar
```
Comment (JoshRosen) [inline dev/deps/spark-deps-hadoop-2.2]: Thanks to this shading, we no longer need to add exclusions for Kryo's ASM dependencies.

**Comment by rxin** [comment]
No API change?

**Comment by JoshRosen** [comment]
Seemingly none that affect us so far. There are some binary incompatibilities in `KryoPool` (it became an interface), but that's not going to impact Spark itself.

**Comment by JoshRosen** [comment]
Test failures appear unrelated.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Huh, weird. It looks like this somehow _is_ a legitimate failure. I'll investigate.

**Comment by johnynek** [comment]
quick glance, it looked to me like the failures were hive related. I don't know if you use any hive code, but could the fact that hive also uses kryo create a diamond that gives a runtime failure?

Just a shot in the dark.

**Comment by magro** [comment]
Btw, we'll have a kryo 3.1.0 release soon (depends on pending fixes regarding handling of generics). This will contain an incompatibility regarding Closure handling (https://github.com/EsotericSoftware/kryo/pull/415), not sure if that's relevant for you.

**Comment by JoshRosen** [comment]
Yep, I think that this is due to conflicts with Hive's version of Kryo. @vanzin, @steveloughran: Given that our execution Hive (1.2.1) uses Kryo 2.x, do we need to upgrade to Hive 2 in order to use a Kryo 3.x-compatible version of Hive ([HIVE-12175](https://issues.apache.org/jira/browse/HIVE-12175))?

**Comment by vanzin** [comment]
I haven't looked at the failures here, but the upstream version of Hive shades Kryo, so hopefully it wouldn't be affected. But I don't know whether Steve made any changes to Spark's fork.

**Comment by JoshRosen** [comment]
Here's the branch that our custom `org.spark-project.hive` artifacts seem to have been published from: https://github.com/apache/hive/compare/branch-1.2...steveloughran:stevel/branch-1.2.1.spark

Looking at the changes there, it appears that our custom Hive 1.2.1 downgrades Kryo from 2.22 to 2.21. However, it appears that our custom build undoes the shading of Kryo and adds shading of Protobuf.

**Comment by JoshRosen** [comment]
I understand why we needed to add shading of Protobuf, but why did we need to remove shading of Kryo? Since I suspect that we do not want to couple the upgrading to Kryo 3 with an upgrade to Hive 2.0, would it make sense to publish a `org.spark-project.hive` 1.2.1-0 (or similar) release to re-add that shading of Kryo?

**Comment by vanzin** [comment]
It should be possible to leave Kryo shaded, but it will require some code changes. We do that in CDH, here's our diff: https://github.com/cloudera/spark/commit/0d8339b6f45460bba67a53ca88d423a301ab98c3#diff-57959d009ae01f8799a2c4d461e96a76

I think that was the only kryo-related change, although I made that change a while ago and might have forgotten other places.

**Comment by rxin** [comment]
how is kryo used in hive? i suspect we might no longer be using that in spark 2.0, once we get rid of all the ddls.

**Comment by steveloughran** [comment]
1. Hive uses Kryo "the guava of serialization" internally; I don't know the specifics, but its not insignificant.
1. they moved ahead of spark's version to fix some bugs; that's the kryo which repackaged things. 
1. Kryo types were used across methods called from Spark; the unshading was needed to get things to link, same with the switch from 2.22 to 2.21 (and a change to some files). It was less traumatic to push Hive back slightly than to try to force an update into unknown territory for chill & dependencies, even if it were possible.

Upgrading to hive 2 would finally get everything in sync: ideally eliminate the need to have a custom hive JAR at all. Essentially all spark needs is a version of Hive with parameters that can be exchanged across all needed methods, with the conflict packages shaded, and the non-conflict packages generally omitted

If you want to find out where things don't link, edit the spark pom to pull in org.apache.hive/hive and see what breaks.


(of course, the other thing that would be nice would be for Hive to make it possible to subclass their thrift service cleanly; I've got a PoC of that somewhere ... it's not hard to add)

**Comment by steveloughran** [comment]
...Thinking about this; it might be possible to go to hive with a shaded kryo, with the invocation of those methods which exchange kryo types referring to the shaded values. That doesn't do anything for marshalling stuff between kryo versions

**Comment by JoshRosen** [comment]
Grepping `sql/hive` for `kryo` turns up one explicit usage in HiveShim's `HivefunctionWrapper` where it looks like we take an instance of `Kryo` provided by Hive and use that instance to deserialize things: https://github.com/apache/spark/blob/8ba2b7f28fee39c4839e5ea125bd25f5091a3a1e/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala#L171. Assuming that we use a shaded Kryo in Hive, I think we could just import the shaded Kryo here and use that import in our `[de]serializeObjectByKryo` methods' signatures.

I don't spot any other Hive-specific uses of Kryo which look problematic, so it seems plausible that we can shade without major problems. Am I overlooking other uses of Kryo to exchange data between Spark and Hive internals?

**Comment by steveloughran** [comment]
I don't recall doing any reflection related stuff to work with kryo; problems were showing up in compilation.

There's one more thing to worry about, hive minimal vs uber jars. It may still be necessary to have a custom 1.2.x JAR, just to keep all the other cruft out.

**Comment by JoshRosen** [comment]
My understanding so far:

- [Marcelo showed](https://github.com/apache/spark/pull/12076#issuecomment-204129446) that we can use shaded Kryo in Hive.
- While the Protobuf shading might have originally been introduced for Hadoop 1.x support, it's probably a good idea to keep the shading because Protobuf is a conflict-prone dependency.
- We can't use the regular `hive-exec` artifacts because they package tons of dependencies without relocation.
- We can't use the `core`-classified `hive-exec` 1.2.1 artifacts because they don't shade the things that we need shaded. Even if we could use this artifact, it is a pain to consume because its POM does not declare required dependencies, so we have to manually add all of Hive's transitive deps as direct deps. We've already sort of done this, but I believe that it's unnecessary and am in the process of undoing a lot of those changes as part of #1217.

Therefore, I think there's no avoiding having to publish another custom Hive 1.2.1 build. I propose that we do so by editing the 1.2.1-spark POM to restore Kryo shading. I'll work on this tomorrow and loop back to this discussion once I've completed the required Hive dep. bumps.

**Comment by steveloughran** [comment]
Makes sense. Note that getting the SBT dependencies to match maven's was a complete nightmare; it'd probably have been easier to write a new version reconciler for ivy than solve it for spark alone. You shouldn't have to do much to that listing as it stands today, but updating to later hive versions will probably hurt again.

FWIW, Hadoop 2.x isn't going to increment protobuf, I'm confident of that to the extent that there are enough people who would veto any attempt. Worry more about guava and jackson

**Comment by JoshRosen** [comment]
Splitting the Hive Kryo shading into a new JIRA: https://issues.apache.org/jira/browse/SPARK-14435

**Comment by JoshRosen** [comment]
The Hive conflict should now be fixed now that #12215 has been merged, so I've updated this and am planning to merge as soon as tests pass.

**Comment by JoshRosen** [comment]
Alright, I'm going to go ahead and merge this. We'll upgrade to 2.12-friendly versions of Chill + Kryo as part of the 2.12 testing patch which will come out later.

**Pull request closed without merging**